modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxRecoveryWithConcurrentRollbackTest.java

```diff
@@ -258,6 +263,120 @@ else if (g1Keys.contains(key))
+            final IgniteEx grid0 = startGrid(0);
+
+            final IgniteEx grid1 = startGrid(1, (UnaryOperator<IgniteConfiguration>)cfg -> cfg
+                .setSystemThreadPoolSize(1).setStripedPoolSize(1));
+
+            final IgniteEx grid2 = startGrid(2);
+
+            grid0.cluster().state(ACTIVE);
+
+            final IgniteCache<Object, Object> cache = grid2.cache(DEFAULT_CACHE_NAME);
+
+            final Transaction tx = grid2.transactions().txStart(PESSIMISTIC, REPEATABLE_READ);
+
+            final Integer g2Key = primaryKeys(cache, 1, 0).get(0);
+
+            cache.put(g2Key, Boolean.TRUE);
+
+            final TransactionProxyImpl<?, ?> p = (TransactionProxyImpl<?, ?>)tx;
+
+            p.tx().prepare(true);
+
+            final List<IgniteInternalTx> txs0 = txs(grid0);
+            final List<IgniteInternalTx> txs1 = txs(grid1);
+            final List<IgniteInternalTx> txs2 = txs(grid2);
+
+            assertTrue(txs0.size() == 1);
+            assertTrue(txs1.size() == 1);
+            assertTrue(txs2.size() == 1);
+
+            final CountDownLatch grid1NodeLeftEventLatch = new CountDownLatch(1);
+
+            grid1.events().localListen(new PE() {
+                @Override public boolean apply(Event evt) {
+                    grid1NodeLeftEventLatch.countDown();
+
+                    return true;
+                }
+            }, EventType.EVT_NODE_LEFT);
+
+            final CountDownLatch grid1BlockLatch = new CountDownLatch(1);
+
+            // Block recovery procedure processing on grid1.
+            grid1.context().pools().getSystemExecutorService().execute(() -> U.awaitQuiet(grid1BlockLatch));
+
+            final int stripe = U.safeAbs(p.tx().xidVersion().hashCode());
+
+            // Block stripe tx recovery request processing on grid1.
+            grid1.context().pools().getStripedExecutorService().execute(stripe, () -> U.awaitQuiet(grid1BlockLatch));
+
+            // Prevent finish request processing on grid0.
+            spi(grid2).blockMessages(GridDhtTxFinishRequest.class, grid0.name());
+
+            // Prevent finish request processing on grid1.
+            spi(grid2).blockMessages(GridDhtTxFinishRequest.class, grid1.name());
+
+            runAsync(() -> {
+                grid2.close();
+
+                return null;
+            });
+
+            try {
+                tx.close();
+            }
+            catch (Exception ignored) {
+                // Don't bother if the transaction close throws in case grid2 appear to be stopping or stopped already
+                // for this thread.
+            }
+
+            // Wait until grid1 node detects primary node left.
+            grid1NodeLeftEventLatch.await();
+
+            // Wait until grid1 receives the tx recovery request and the corresponding processing task is added into
+            // the queue.
+            assertTrue("tx recovery request received on grid1", GridTestUtils.waitForCondition(() -> grid1.context()
```

Comment:
assert message is about situation when assert failed

Reply:
Fixed.